authui/src/form.ts

```diff
@@ -1,25 +1,150 @@
+import { visit, clearCache } from "@hotwired/turbo";
 import { Controller } from "@hotwired/stimulus";
+import axios, { Method } from "axios";
+import {
+  disableAllButtons,
+  hideProgressBar,
+  showProgressBar,
+  progressEventHandler,
+} from "./loading";
+import { handleAxiosError } from "./messageBar";
 
-// We used to disable buttons during form submission.
-// It is now supported by Turbo natively.
-// See https://turbo.hotwired.dev/handbook/drive#form-submissions
-export class TurboformController extends Controller {
-  submitEnd(e: CustomEvent) {
-    const submitter: HTMLElement | undefined =
-      e.detail.formSubmission.submitter;
-    const formElement: HTMLFormElement = e.detail.formSubmission.formElement;
-    const location: URL | undefined =
-      e.detail.formSubmission.result?.fetchResponse.location;
-    if (location != null) {
-      const action = location.searchParams.get("x_turbo_action");
-      if (action != null && action !== "") {
-        if (submitter != null) {
-          submitter.setAttribute("data-turbo-action", action);
-        } else {
-          formElement.setAttribute("data-turbo-action", action);
-        }
+// Turbo has builtin support for form submission.
+// We once migrated to use it.
+// However, redirect to external is broken because
+// the redirect is made with fetch, which is subject to CORS.
+// A typical problem is the support for the post login redirect URI.
+// The post login redirect URI is usually an external link that
+// the origin of the link does not list our origin as allowed origin.
+// If we rely on Turbo to handle form submission,
+// the post login redirect URI will be broken.
+// Therefore, we reverted to handle form submission ourselves.
+// To disable the builtin form submission of Turbo,
+// I studied its source code and discovered that
+// Turbo checked in the bubble event listener to see if
+// the event is prevented.
+// So I added a capture event listener to call preventDefault()
+// to stop Turbo from submitting forms.
+//
+// See https://github.com/authgear/authgear-server/issues/2333
+export class XHRSubmitFormController extends Controller {
+  revertDisabledButtons: { (): void } | null = null;
+  forms: HTMLFormElement[] = [];
+
+  // Revert disabled buttons before Turbo caches the page
+  // To avoid flickering in the UI
+  beforeCache = () => {
+    if (this.revertDisabledButtons) {
+      this.revertDisabledButtons();
+    }
+  };
+
+  onSubmitCapture = (e: Event) => {
+    e.preventDefault();
+  };
+
+  onSubmit = (e: Event) => {
+    this.submitForm(e);
+  };
+
+  async submitForm(e: Event) {
+    const form = e.currentTarget as HTMLFormElement;
+
+    if (form.querySelector('[data-turbo="false"]')) {
+      return;
+    }
+
+    // Do not stop propagation so that GTM can recognize the event as Form Submission trigger.
```

Comment:
In the original `XHRSubmitFormController` implementation, we had the following code here. Do we need to add it back?
```
if (e.defaultPrevented) {
      return;
}
e.preventDefault();
```

Reply:
Intentionally removed because we always call preventDefault() in the capture event listener. If we were to keep it, we always early return.